File: 刷题/Hot 100/146. LRU 缓存.go
type Node struct {
    key, value int
    pre *Node
    next *Node
}

type LRUCache struct {
    cache map[int]*Node
    capacity int
    dummy *Node
}

func NewNode(key, value int) *Node {
    node := new(Node)
    node.key = key
    node.value = value
    node.pre = node
    node.next = node
    return node
}

func Constructor(capacity int) LRUCache {
    lru := LRUCache{
        capacity: capacity,
        cache: make(map[int]*Node),
        dummy: NewNode(0, 0),
    }
    return lru
}

func (this *LRUCache) remove(node *Node) {
    node.pre.next=node.next
    node.next.pre=node.pre
}

func (this *LRUCache) addToFront(node *Node) {
    node.pre = this.dummy
    node.next = this.dummy.next
    this.dummy.next.pre = node
    this.dummy.next = node
}

func (this *LRUCache) get_node(key int) *Node{
    if node, ok := this.cache[key]; ok {
        this.remove(node)
        this.addToFront(node)
        return node
    }
    return nil
}

func (this *LRUCache) Get(key int) int {
    node := this.get_node(key)
    if node == nil {
        return -1
    }
    return node.value
}


func (this *LRUCache) Put(key int, value int)  {
    node := this.get_node(key)
    if node != nil {
        node.value = value
        return
    }
    
    node = NewNode(key, value)
    this.cache[key] = node
    this.addToFront(node)
    if len(this.cache) > this.capacity {
        tail := this.dummy.pre
        delete(this.cache, tail.key)
        this.remove(tail)
    }
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: 刷题/Hot 100/146. LRU 缓存.cpp
class Node{
public:
    int key, value;
    Node* pre;
    Node* next;
    Node(int _key=0, int _value=0): key(_key), value(_value){}
};

class LRUCache {
private:
    unordered_map<int, Node*> cache;
    int capacity;
    Node* dummy;
public:
    LRUCache(int _capacity):capacity(_capacity), dummy(new Node()) {
        dummy->pre=dummy;
        dummy->next=dummy;
    }

    void remove(Node* node) {
        node->pre->next=node->next;
        node->next->pre=node->pre;
    }

    void addToFront(Node* node) {
        node->pre=dummy;
        node->next=dummy->next;
        dummy->next->pre=node;
        dummy->next=node;
    }

    Node* get_node(int key) {
        auto it = cache.find(key);
        if (it == cache.end()) {
            return NULL;
        }
        auto node = it->second;
        remove(node);
        addToFront(node);
        return node;
    }

    int get(int key) {
        Node *node = get_node(key);
        if(node != NULL) {
            return node->value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        Node* node = get_node(key);
        if(node!=NULL) {
            node->value = value;
            return;
        }
        cache[key] = node = new Node(key, value);
        addToFront(node);
        if(cache.size() > capacity) {
            Node* tail_node = dummy->pre;
            cache.erase(tail_node->key);
            remove(tail_node);
            delete(tail_node);
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */